### PR TITLE
[Refactor] Introduce `RecordBase` to share common fields between `Record` and `RecRecord` terms

### PR DESCRIPTION
**Pull request by matthew-healy**
### Why is this needed?

As part of refactoring record contracts with polymorphic tails, we're going to introduce an `Option<Box<Term>>` field to both `Record`s and `RecRecord`s which will hold the sealed portion of records passed through contracts with polymorphic tails (see [this issue](https://github.com/tweag/nickel/issues/201) for more info, in particular the section "Coarse sealing built-in"). 

To make this simpler, this PR groups together the existing common fields into a `RecordBase` struct. It also introduces two functions on this type to avoid the need for consumers to destructure & then rebuild `RecordBase`s everywhere. 

### Notes

- Both `map_fields` and `filter_map_fields` take their function parameter as `mut`. This is because in certain places we build up additional state while iterating (e.g. by pushing the `Ident`s into a `Vec`). I _think_ it's reasonable to do this, but would appreciate other opinions in case there's something I missed.

### Recommendation for reviewers

Each commit is self-contained & passes the whole test suite. Several also have extra detail in the commit messages. As such I'd recommend reviewing commit-by-commit rather than trying to review the whole PR as a single unit.

### CODE REVIEW

**Review by vkleen — APPROVED** [review]
The changes look good to me. Another idea for the name of `Record` could be `RecordBase`; it's not a record by itself, but it's the base for both kinds of records.

**Comment by matthew-healy** [comment]
@vkleen Huh, I hadn't considered `RecordBase` but I prefer it to `BaseRecord` for reasons I can't quite put into words. Unless I hear an alternative suggestion I'll make that change later today. Thank you! :pray:

File: src/term.rs
Comment (sir4ur0n) [inline src/term.rs:429]:
```suggestion
    /// Used to group together fields common to both the [Record](Term::Record) and
    /// [RecRecord](Term::RecRecord) terms.
```

File: src/typecheck/eq.rs
Comment (sir4ur0n) [inline src/typecheck/eq.rs:271]: Same remark on nested pattern matching

File: src/typecheck/mod.rs
Comment (sir4ur0n) [inline src/typecheck/mod.rs:153]: Why do we need `&` now?

File: src/eval/merge.rs
Comment (matthew-healy) [reply src/eval/merge.rs:365]: Yes, we theoretically could, using record syntax, e.g.:

```rust
(Term::Record(RecordBase { fields: mut m1, attrs: attrs1 }, Term::Record(RecordBase { fields: mut m2, attrs: attrs2 }) => ...
```

But as you can see, the result is significantly more verbose because you need to specify which field names map to which new bindings, since record destructuring isn't based on position. 

Ultimately I think binding to the top-level struct & then accessing its fields is probably simpler.

File: src/term.rs
Comment (matthew-healy) [reply src/term.rs:454]: The main benefit is that it leaks less information about the internal structure of the struct to consumers of this type. We can freely add fields to the struct without ever changing the callsites of `with_fields` or `empty` unless they have a valid reason to change. I'd also argue that `RecordBase::empty()` is a lot more explicit about its intention than `RecordBase::with(None, None)`.

I think it's generally idiomatic in Rust to have a "main" `new` constructor, and then custom versions for specific use-cases. The additional constructors generally take fewer arguments, with the remaining stored values set to sensible defaults (often, but not always, using the `Default` trait as we do here). 

As a result, if I saw a callsite for the `with` constructor you suggest, I'd expect  `RecordBase` to look something like this:

```rust
struct RecordBase {
    fields: Option<HashMap<Ident, RichTerm>>,
    attrs: Option<RecordAttrs>,
}
```

I wouldn't necessarily expect that passing a `None` value for either argument would result in a `Default` value of the type argument being used instead.

File: src/term.rs
Comment (matthew-healy) [reply src/term.rs:461]: Good catch! I'll fix this before merging.

File: src/typecheck/mod.rs
Comment (matthew-healy) [reply src/typecheck/mod.rs:153]: I was also initially a little surprised this was required, but essentially: accessing a field of a borrowed struct implicitly deferences that struct in order to find the field, so if `x: &X` then `x.y` actually results in `(*x).y`.

That is to say that if we have:

```rust
struct MyStruct {
    x: String
}

let my_struct = MyStruct { x: "hello".into() };
let borrowed = &my_struct;
let val = borrowed.x;
```

Then `val` would have type `String` - i.e. it's not borrowed. This example would pas type checking, but would fail the borrow checker as it attempts to move the value of `x` out of a borrowed struct. If we instead wrote `let val = &borrowed.x;` then we'd get a reference & the borrow checker would be happy.

Previously we had `&HashMap<_, _>`, so no explicit borrow was required, but now that we're accessing a field of a struct, we need to make it explicit that we're not trying to move that field.

File: src/eval/merge.rs
Comment (sir4ur0n) [reply src/eval/merge.rs:365]: > It also avoids the need to change this part of the code when adding a new field to RecordBase

I see that as a benefit rather than a problem because some call sites may need to be updated as well. But in any case I won't block on that

File: src/term.rs
Comment (sir4ur0n) [inline src/term.rs:448]: Rather than having an ad hoc function, can we/should we have `Record` implement the [Default](https://doc.rust-lang.org/std/default/trait.Default.html) trait?

File: src/term.rs
Comment (sir4ur0n) [reply src/term.rs:461]: Just to save another back-and-forth: this comment also applies to `filter_update_fields` :wink:

File: src/typecheck/mod.rs
Comment (sir4ur0n) [reply src/typecheck/mod.rs:153]: Thank you for the explanation!

File: src/term.rs
Comment (yannham) [reply src/term.rs:448]: It is customary to have a default implementation indeed. You can probably just derive it for `Record` using `#[derive(Default)]`, because the subpart implements `Default`, I think. Then it's standard to add a `new()` method which is just an alias for `default()`.

File: src/eval/merge.rs
Comment (yannham) [reply src/eval/merge.rs:365]: Both approaches have arguments, but one appeal of not doing the destructuring is to declare and use fewer top-level identifiers. If you're giving a name to each subfield anyway `m1, m2, attrs1, attrs2`, IMHO you may as well use destructuring, because as Julien brings up, the soundness of the implementation is more resilient to refactoring.

Comment (matthew-healy) [reply src/eval/merge.rs:365]: I'm not entirely sure what you mean by "soundness" in this context. As far as I can tell, introducing a new field is a purely additive change that can't break existing behaviour. If the new field is _intended_ to change existing behaviour, then I'd expect a change in the associated tests to go alongside it, which would prompt changes here. All destructuring here would do is require you to add either `, new_field: _` or `, ..` to the destructuring, and if you choose the latter then you're in the same state as you are here. It's entirely possible I'm missing something though.

Either way, I'm glad you mentioned the top-level identifiers, as I actually meant to remove those - they were only there to get things compiling faster. :sweat_smile: I'll do that as part of the updates I'm making now.

**Review by yannham — APPROVED** [review]
Suggestion (that I saw somewhere in rustc, rust-analyzer, or another language implem): call the thing that goes int a `Record` variant, `RecordData`.

**Comment by matthew-healy** [comment]
@yannham Good suggestion! I've also seen the `_Data` naming convention somewhere, but completely forgot about it until you mentioned it. I think in this case I'll stick with `RecordBase` because this represents part of two variants, but we can always come back & rename it later.

File: src/term.rs
Comment (yannham) [reply src/term.rs]: Hmm, you're right, although we already mix static and runtime constructs (for example ops have an `eval_mode()` function that returns `eval::EvalMode`), it's not a great thing to do in general. That being said, we can still implement this function as a standalone helper somewhere inside the evaluation module (and get rid of `filter_update_fields`, if it's not used anymore) ?

Another solution is to use the trait `RecordBaseExt` pattern (use a boilerplate trait to add a function to `RecordBase` from within `eval`). But I think it's not really required here, because we don't need to access private data.

**Comment by yannham** [comment]
`RecordBase` is fine, but I think the `Base` part doesn't convey much information. `RecordFields`? `RecordMap`? Those are just suggestion, you can stick to `RecordBase` if you think this is better.

Another random thought: have you thought about inlining `RecordAttrs` into `RecordBase`? I think I originally put it in a separate variant for the very same reason (for it to be shared between `RecRecord` and `Record`, and also to separate it from the rest of the data). I think going either way is fine (separated or inlined), I'm just wondering.

File: src/term.rs
Comment (matthew-healy) [reply src/term.rs]: I guess the only awkwardness this creates is that now both `map_fields` and the new standalone function have broadly the same implementation, but can't be deduplicated (i.e. `map_fields` can no longer call out to `filter_map_fields` for its implementation). But that's still a lot less duplication than before, so I guess I'll go with that. :grin:

**Comment by matthew-healy** [comment]
Following the discussions outside of this PR, I've decided to:

- rename `RecordBase` to `RecordData`
- inline `RecordAttrs` into `RecordData`

I'll do that now, then merge once it passes CI.

**Comment by matthew-healy** [comment]
Actually, I'd like to close this PR & then inline `RecordAttrs` separately, as I think that change will lead to some non-trivial changes around how records are merged.

**Pull request merged**
